### ISSUE TITLE
HTTP: cancel_http_readonly_queries_on_client_close=1 fails with 19.14

### ISSUE THREAD

**Issue by geldot** [ClickHouse/ClickHouse#7019]
With the 19.14 series this hangs:

```
curl 'http://localhost:8123/?query=SELECT%201&cancel_http_readonly_queries_on_client_close=1'
```

When the `cancel_http_readonly_queries_on_client_close` setting is omitted or set to `0`, it works.

With the 19.13 series and earlier the above request successfully returns results as expected. (We did not check if it correctly cancelled the requests but assume so.)

The `cancel_http_readonly_queries_on_client_close` feature is documented as: "Cancel HTTP readonly queries when a client closes the connection without waiting for response."

I haven't investigated why test `00834_cancel_http_readonly_queries_on_client_close.sh` passes but it looks like that test may be flawed or have missing coverage.

**Comment by geldot** [comment]
OK, it looks like missing coverage: The existing test is only checking for a timeout, which also happens when the HTTP request hangs, so it doesn't fail). A more basic `SELECT 1` test as above will catch this.

**Comment by geldot** [comment]
The regression has now been introduced to the 19.13 series as of yesterday's point release:

HTTP requests with `cancel_http_readonly_queries_on_client_close=1`:

Failing: 19.13.5.44
Working: 19.13.4.32

**Comment by geldot** [comment]
Looks like one of these two Poco updates (#6386 and #6466) could be responsible. Didn't have a chance to build and test yet

**Comment by silviucpp** [comment]
For me on master works ok. When I close the connection the query is canceled and I can see in logs:

```
DB::Exception: Query was cancelled (version 19.15.1.1) 
```

BTW if you send the requests via POST you need to specify readonly=2 as well in the query params.

**Comment by silviucpp** [comment]
Only problem I can see is with CREATING TEMPORARY TABLES.

Based on docs: "Temporary tables disappear when the session ends, including if the connection is lost."

But if I run a query like `CREATE TEMPORARY TABLE `temp1` AS SELECT number FROM system.numbers FORMAT JSON` no matter If I close the connection the query will run on the server till runs out of memory.

@alexey-milovidov is this expected ?

**Comment by silviucpp** [comment]
@alexey-milovidov do you have a bit of time to clarify my question ? is this expected or it's something that needs to be fixed ?

**Comment by alexey-milovidov** [comment]
> @alexey-milovidov is this expected?

No. The query should be cancelled if the server can detect that the client is disconnected.

**Comment by silviucpp** [comment]
Problem is related to https://github.com/ClickHouse/ClickHouse/issues/7736 and I described there from where it comes.

**Comment by citrin** [comment]
I see the same problem with 19.16.14.65.
Query hangs until connection is closed by a HTTP client (even when HTTP timeout is much higher than `max_execution_time` settings).

**Comment by champtar** [comment]
@filimonov you can add some more labels, this is still broken in latest stable/lts releases
```
$ podman run -d --network host --name clickhouse-server yandex/clickhouse-server:20.3.8.53
$ curl 'http://127.0.0.1:8123/?query=SELECT%201' -u 'default:' -m2
1
$ curl 'http://127.0.0.1:8123/?cancel_http_readonly_queries_on_client_close=true&query=SELECT%201' -u 'default:' -m2
curl: (28) Operation timed out after 2001 milliseconds with 0 bytes received
```

**Comment by champtar** [comment]
This is still broken in latest lts (20.3.9), it should be in 20.3.10 (https://github.com/ClickHouse/ClickHouse/pull/10684#issuecomment-624912487)
```
$ podman run -d --network host --name clickhouse-server yandex/clickhouse-server:20.3.9.70
$ curl 'http://127.0.0.1:8123/?cancel_http_readonly_queries_on_client_close=true&query=SELECT%201' -u 'default:' -m2
curl: (28) Operation timed out after 2001 milliseconds with 0 bytes received
```
It is fixed in 20.4.2.9
```
$ podman run -d --network host --name clickhouse-server yandex/clickhouse-server:20.4.2.9
$ curl 'http://127.0.0.1:8123/?query=SELECT%201' -u 'default:' -m2
1
$ curl 'http://127.0.0.1:8123/?cancel_http_readonly_queries_on_client_close=true&query=SELECT%201' -u 'default:' -m2
1
```
